File: src/main/java/com/demonwav/mcdev/platform/hybrid/SpongeForgeProjectConfiguration.java
package com.demonwav.mcdev.platform.hybrid;

import com.demonwav.mcdev.buildsystem.BuildSystem;
import com.demonwav.mcdev.platform.PlatformType;
import com.demonwav.mcdev.platform.forge.ForgeProjectConfiguration;
import com.demonwav.mcdev.platform.sponge.SpongeProjectConfiguration;
import com.demonwav.mcdev.platform.sponge.SpongeTemplate;
import com.demonwav.mcdev.util.Util;

import com.intellij.ide.util.EditorHelper;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiJavaFile;
import com.intellij.psi.PsiManager;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class SpongeForgeProjectConfiguration extends ForgeProjectConfiguration {

    public boolean generateDocumentation = false;
    public String spongeApiVersion;

    public SpongeForgeProjectConfiguration() {
        // We set our platform type to sponge because we want it to provide us the dependency. The GradleBuildSystem
        // will properly handle us as a combined project
        type = PlatformType.SPONGE;
    }

    @Override
    public void create(@NotNull Project project, @NotNull BuildSystem buildSystem, @NotNull ProgressIndicator indicator) {
        Util.runWriteTask(() -> {
            try {
                indicator.setText("Writing main class");
                VirtualFile file = buildSystem.getSourceDirectories().get(0);
                String[] files = this.mainClass.split("\\.");
                String className = files[files.length - 1];
                String packageName = this.mainClass.substring(0, this.mainClass.length() - className.length() - 1);
                file = getMainClassDirectory(files, file);

                VirtualFile mainClassFile = file.findOrCreateChildData(this, className + ".java");
                SpongeTemplate.applyMainClassTemplate(
                        project,
                        mainClassFile,
                        packageName,
                        className,
                        hasDependencies(),
                        generateDocumentation
                );

                writeMcmodInfo(project, buildSystem);

                PsiJavaFile mainClassPsi = (PsiJavaFile) PsiManager.getInstance(project).findFile(mainClassFile);
                if (mainClassPsi == null) {
                    return;
                }
                PsiClass psiClass = mainClassPsi.getClasses()[0];

                SpongeProjectConfiguration.writeMainSpongeClass(
                        project,
                        mainClassPsi,
                        psiClass,
                        buildSystem,
                        pluginName,
                        description,
                        website,
                        hasAuthors(),
                        authors,
                        hasDependencies(),
                        dependencies
                );

                // Set the editor focus on the main class
                EditorHelper.openInEditor(mainClassPsi);
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/hybrid/SpongeForgeProjectConfiguration.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2018 minecraft-dev
 *
 * MIT License
 */

@file:Suppress("Duplicates")

package com.demonwav.mcdev.platform.hybrid

import com.demonwav.mcdev.buildsystem.BuildSystem
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.forge.ForgeProjectConfiguration
import com.demonwav.mcdev.platform.sponge.SpongeTemplate
import com.demonwav.mcdev.platform.sponge.writeMainSpongeClass
import com.demonwav.mcdev.util.runWriteTask
import com.intellij.ide.util.EditorHelper
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiJavaFile
import com.intellij.psi.PsiManager

class SpongeForgeProjectConfiguration : ForgeProjectConfiguration() {

    override var type: PlatformType = PlatformType.SPONGE

    var spongeApiVersion = ""

    override fun create(project: Project, buildSystem: BuildSystem, indicator: ProgressIndicator) {
        if (project.isDisposed) {
            return
        }

        val baseConfig = base ?: return
        val dirs = buildSystem.directories ?: return

        runWriteTask {
            indicator.text = "Writing main class"
            var file = dirs.sourceDirectory
            val files = baseConfig.mainClass.split(".").toTypedArray()
            val className = files.last()
            val packageName = baseConfig.mainClass.substring(0, baseConfig.mainClass.length - className.length - 1)
            file = getMainClassDirectory(files, file)

            val mainClassFile = file.findOrCreateChildData(this, className + ".java")
            SpongeTemplate.applyMainClassTemplate(
                project,
                mainClassFile,
                packageName,
                className,
                hasDependencies()
            )

            writeMcmodInfo(project, baseConfig, buildSystem, dirs)

            val mainClassPsi = PsiManager.getInstance(project).findFile(mainClassFile) as? PsiJavaFile ?: return@runWriteTask
            val psiClass = mainClassPsi.classes[0]

            writeMainSpongeClass(
                project,
                mainClassPsi,
                psiClass,
                buildSystem,
                baseConfig.pluginName,
                baseConfig.description ?: "",
                baseConfig.website ?: "",
                hasAuthors(),
                baseConfig.authors,
                hasDependencies(),
                dependencies
            )

            // Set the editor focus on the main class
            EditorHelper.openInEditor(mainClassPsi)
        }
    }
}
